### ISSUE TITLE
Decide on approach for exceptions thrown by public methods

### ISSUE THREAD

**Issue by mistermoe** [TBD54566975/web5-kt#144]
We should decide on a convention we want to follow for exceptions thrown by public methods. 

Proposal:
* avoid explicitly throwing `Exception` and `RuntimeException`. Both are far too general, we should throw exceptions that indicate a specific error case. We're using `Exception` and `RuntimeException` in a handful of places at the moment. [example](https://github.com/TBD54566975/web5-kt/blob/c9bfb61ace04cb6a8f6afcbda6d148820c63e4ad/credentials/src/main/kotlin/web5/sdk/credentials/StatusListCredential.kt#L232)
* favor `Exception` over `Error` for custom exception classes. Conventionally, `*Error` are not expected to be caught. They are conditions that typically affect the JVM's ability to continue execution. Examples include `OutOfMemoryError` and `StackOverflowError`. `Error` is intended to indicate a serious problem. Example of us using the word `Error` [here](https://github.com/TBD54566975/web5-kt/blob/a644e6f129cabc85906099d8e79497a1d18fff71/credentials/src/main/kotlin/web5/sdk/credentials/PresentationExchange.kt#L74)
* ensure that exceptions thrown from our public api surface are either included as part of java/kotlin _or_ are custom exception classes that come directly from our package as opposed to third party dependencies

We should decide on our approach here and then prioritize implementing the approach before our codebase grows larger.

**Comment by andresuribe87** [comment]
Super important topic, thanks for raising @mistermoe !

I agree with the first two points. 

Re: to 

> ensure that exceptions thrown from our public api surface are either included as part of java/kotlin or are custom exception classes that come directly from our package as opposed to third party dependencies

How do you do this in practice? For instance, if a dependency throws an custom exception, how can we guarantee that we'll not throw it? We would presumably need to surround most external dependency calls with `try/catch` blocks, but that's difficult to maintains because: (a) makes code incredibly verbose and (b) it's hard to enforce in PRs.

**Comment by mistermoe** [comment]
great point @andresuribe87 . i was feeling 😅 🙈  as i typed that one out. 

Typically, what i've done is look at the javadoc/kdoc for method i'm calling to see if the exceptions thrown are documented. if they aren't documented, i click into the implementation to see if custom exceptions are thrown, but this can easily become a blackhole if the method i'm inspecting calls n other methods from other third party dependencies.

you're right, it does seem hard if not impossible to enforce in PRs. The motivation behind that bullet point is primarily two reasons:
* if we're throwing custom exceptions from third party dependencies, do we have to declare those dependencies as `api` instead of `implementation` so that consumers can actually import the exception?
* i think it'd be beneficial for us to know exactly what exceptions are expected from our public api surface. 

all that said,  should we choose to adopt that last bullet, i suppose it'd be more of a recommendation, and something we retroactively account for when/if we encounter third party custom exceptions bubbling up. Though it's far less of a concern so long as consumers can actually import the custom exception

**Comment by andresuribe87** [comment]
> if we're throwing custom exceptions from third party dependencies, do we have to declare those dependencies as api instead of implementation so that consumers can actually import the exception?

I think the answer here is that it depends. If we're suggesting that the consumers can do something to handle the exception, then I would say they need to be part of the public API. Otherwise, then hiding it behind the implementation makes sense to me.

> i think it'd be beneficial for us to know exactly what exceptions are expected from our public api surface.

I agree! And I think it goes hand in hand with the answer above. One guideline that we could decide on is:
* Any expected exceptions MUST be thrown by public API methods. Guidance: ask yourself, can a client reasonably be expected to programmatically recover when the exception is thrown? 
  * Consumers MUST be able to import expected exceptions without changing their gradle files.
* All expected exceptions thrown by a public method MUST be marked with the `@Throws` annotation.
* All exceptions marked with the `@Throws` annotation MUST be tested.

The first bullet point is left as a judgement call (what is an expected exception?) to the author and reviewer of the PR. I don't think there are hard and fast rules here.

Thoughts?

**Comment by mistermoe** [comment]
that makes sense to me!

**Comment by KendallWeihe** [comment]
Really good work from @jiyoontbd on this [here](https://github.com/TBD54566975/web5-kt/pull/154)

I want to add some thoughts with respect to the discussion above regarding exceptions from our dependencies. In a perfect world, our programmatic interface should never expose an exception from a transitive dependency; keyword "perfect" which of course is not attainable. The reasoning being, developers are installing "web5" and therefore have the expectation to develop against web5, and not web5 plus some transitive (and likely lower-level) dependency which the web5 devs decided was a good choice.

The way I would frame this up is two-fold:

1. Custom exceptions we define for semantic reasoning
2. "Everything else" 😄 

It's the 'ole adage of, _"just return a 500 for everything you don't know of."_ 

For (1), these exceptions exist for the purpose of driving SDK consumer logic, which ought to be gracefully handled. Exceptions are good for this, rather than return types because exceptions can bubble-up N-number of calls on the call stack.

For (2), these exceptions are pure failure scenarios, which cannot be gracefully handled and inevitably require human input. This is the "catch-all" exception case, for which, we could define a standard `Web5Exception`, which is slightly preferred to the standard Kotlin `Exception` because it adds one additional (albeit small) layer of semantic reasoning for the SDK consumer (basically: "if `Web5Exception` then something went wrong, but else if `Exception` then something **REALLY** went wrong" lol). Within the context of the `Web5Exception` we would include any useful information from the underlying exception, which would help us diagnose issues caused from transitive dependencies.

So basically, what I'm saying is, every public function in the SDK should be wrapped in a try/catch and within the catch we have conditional logic which goes: if this is one of our custom exceptions then throw it, else throw the catch-all `Web5Exception`. Which, is something not so easily enforced during code reviewal (easy to forget), but perhaps a tool like detekt has a rule for it.
